File: Merge_Sorted_array.java
/*
    Approach 1 : using Auxilary Array 

        time complexity : O(NLogN)
        space complexity : O(M+N)
        is worked on leetcode : YES

    Approach 2: using Two pointer
        Time Complexity : O(M+N)
        Space Complexity : O(1)
        is worked on leetcode : YES
*/

import java.util.Arrays;
public class Merge_Sorted_array{

    //1 8 9
    //2 3 4 6 7 
    public void merge_using_two_pointer(int [] nums1, int m, int[] nums2, int n){
    
        int i=m-1;
        int j=n-1;
        int k= m+n-1;
        while(i  >= 0 && j >= 0){
            if(nums2[j] >= nums1[i]){
                nums1[k] =  nums2[j];
                k--;
                j--;
            }
            else{
                nums1[k] = nums1[i];
                i--;
                k--;
            }
        }
        
        // while(i>0){
        //     nums1[k] = nums1[i];
        //     i--;
        //     k--;
        // }
        while(j>= 0){
            nums1[k] = nums2[j];
            j--;
            k--;
        }
    
    }
    
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        int j=0;
        for(int i =m;i< nums1.length;i++){
            nums1[i] =nums2[j];
            j++;
        }
       Arrays.sort(nums1);
       
   }
   public static void main(String[] args) {
       Merge_Sorted_array ob =  new Merge_Sorted_array();
       int[]nums1= {1,8,9,0,0,0,0,0};
       int[]nums2={2,3,4,6,7};
       ob.merge_using_two_pointer(nums1, 3, nums2, 5);
       for(int i=0;i<nums1.length;i++){
           System.out.println(nums1[i]);
       }
   }
}

File: mergeSortedArray.go
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes

/*

Approach:
We use the 2 pointer pattern, but with 3 pointers. Each pointer points to the
end of the 3 arrays - nums1, nums2, and the end of the empty part of nums 1.
We populate the nums1 array from the end as we always know that the last
element of both arrays are the largest in those respective arrays, so we can
just keep comparing whichever is the highest and placing them directly in the
end of nums1.

*/

package main

func merge(nums1 []int, m int, nums2 []int, n int) {
	if len(nums1) == 0 || len(nums2) == 0 {
		return
	}

	p1, p2, p3 := m-1, n-1, m+n-1

	for ; p1 >= 0 && p2 >= 0; p3-- {
		if nums1[p1] > nums2[p2] {
			nums1[p3] = nums1[p1]
			p1--
		} else {
			nums1[p3] = nums2[p2]
			p2--
		}
	}

	for ; p1 >= 0; p1-- {
		nums1[p3] = nums1[p1]
		p3--
	}

	for ; p2 >= 0; p2-- {
		nums1[p3] = nums2[p2]
		p3--
	}
}
